test_conformance/extensions/cl_khr_command_buffer/negative_command_nd_range_kernel.cpp

```diff
@@ -0,0 +1,535 @@
+        cl_int error =
+            clGetDeviceInfo(device, CL_DEVICE_COMMAND_BUFFER_CAPABILITIES_KHR,
+                            sizeof(capabilities), &capabilities, NULL);
+        test_error(error,
+                   "Unable to query CL_DEVICE_COMMAND_BUFFER_CAPABILITIES_KHR");
+
+        bool device_does_not_support_printf =
+            (capabilities & CL_COMMAND_BUFFER_CAPABILITY_KERNEL_PRINTF_KHR)
+            == 0;
+
+        return device_does_not_support_printf;
+    }
+
+    cl_int SetUpKernel() override
+    {
+        cl_int error = CL_SUCCESS;
+
+        const char* kernel_str =
+            R"(
+__kernel void printf_kernel() {
+    printf("Hello World\n");
+  }
+)";
+
+        error = create_single_kernel_helper_create_program(context, &program, 1,
+                                                           &kernel_str);
+        test_error(error, "Failed to create program with source");
+
+        error = clBuildProgram(program, 1, &device, nullptr, nullptr, nullptr);
+        test_error(error, "Failed to build program");
+
+        kernel = clCreateKernel(program, "printf_kernel", &error);
+        test_error(error, "Failed to create printf_kernel kernel");
+
+        return CL_SUCCESS;
+    }
+
+    cl_int SetUp(int elements) override
+    {
+        cl_int error = init_extension_functions();
+        if (error != CL_SUCCESS)
+        {
+            return error;
+        }
+
+        if (elements <= 0)
+        {
+            return CL_INVALID_VALUE;
+        }
+        num_elements = static_cast<size_t>(elements);
+
+        error = SetUpKernel();
+        test_error(error, "SetUpKernel failed");
+
+        command_buffer = clCreateCommandBufferKHR(1, &queue, nullptr, &error);
+        test_error(error, "clCreateCommandBufferKHR failed");
+
+        return CL_SUCCESS;
+    }
+
+    cl_int Run() override
+    {
+        cl_int error = clCommandNDRangeKernelKHR(
+            command_buffer, nullptr, nullptr, kernel, 1, nullptr, &num_elements,
+            nullptr, 0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_OPERATION,
+                               "clCommandNDRangeKernelKHR should return "
+                               "CL_INVALID_OPERATION",
+                               TEST_FAIL);
+
+        return CL_SUCCESS;
+    }
+};
+
+// CL_INVALID_OPERATION if the device associated with command_queue does not
+// support CL_COMMAND_BUFFER_CAPABILITY_DEVICE_SIDE_ENQUEUE_KHR and kernel
+// contains a kernel-enqueue call.
+struct CommandNDRangeKernelWithKernelEnqueueCall : public BasicCommandBufferTest
+{
+    using BasicCommandBufferTest::BasicCommandBufferTest;
+
+    bool Skip() override
+    {
+        bool has_device_enqueue = false;
+        bool cl_version_2_0_or_higher = false;
+
+        OpenCLCFeatures features;
+        get_device_cl_c_features(device, features);
+
+        const Version clc_version = get_device_cl_c_version(device);
```

Comment:
Should use `get_device_latest_cl_c_version` here, that will report a higher version if available, and so is more likely to be able to test this feature.